src/traversals/allsimplepaths.jl

```diff
@@ -0,0 +1,185 @@
+using DataStructures
+
+"""
+    all_simple_paths(g, source, targets, cutoff=nothing)
+
+Returns an iterator that generates all simple paths in the graph `g` from `source` to `targets`.
+If `cutoff` is given, the paths' lengths are limited to equal or less than `cutoff`.
+Note that the length of a path is defined as the number of edges, not the number of elements.
+ex. the path length of `[1, 2, 3]` is two.
+Internally, a DFS algorithm is used to search paths.
+
+# Examples
+
+```jldoctest
+julia> using LightGraphs
+julia> g = complete_graph(4)
+julia> collect(all_simple_paths(g, 1, [4]))
+5-element Array{Array{Int64,1},1}:
+ [1, 4]
+ [1, 3, 4]
+ [1, 3, 2, 4]
+ [1, 2, 4]
+ [1, 2, 3, 4]
+ ```
+"""
+function all_simple_paths(g::AbstractGraph, source::T, targets::Vector{T}; cutoff::Union{Int,Nothing}=nothing) where T <: Integer
```

Comment:
`targets` is a `Vector` here, but in the struct, it's a `Set`. Perhaps we should not type it here, and construct a `Set` out of it in these functions instead of in the inner constructor of the struct.

Reply:
I moved the typecasting from Structure to `all_simple_paths` function. If I misunderstood your point let me know.